## Now let's practice what we have learnt by trying out the following excercise!
Remeber, Google and Stackoverflow are your friends!

### <span style="color: red"> Task 1:</span> The text file you downloaded earlier ('E1_T1.csv') contains ten words. Read the words into a list and use a "for loop" to print out each word. Also print the words as a full sentence (i.e., a single string), as well as the data type of the sentence.

### <span style="color: blue"> Solution:</span>

In [31]:
#importing dependencies
import os
import csv

#getting current directory & stating file name
os.getcwd()
filename = 'data-1/E1_T1.csv'#changes depending on the directory relative to the cwd

#opening file
with open(filename, 'r', newline='') as f:
    #creating reader object
    csv_file = csv.reader(f)
    
    #reading in data
    word_list=next(csv_file)
    
    #printing out each word
    for word in word_list:
        print(word)

    #joining words together into a single string
    joined_words = ' '.join(word_list)
    
    #printing out both the sentence and its data type
    print(joined_words,'\n',f'Data type : {type(joined_words)}')

This
task
is
too
simple
,
I
want
more
challenges
This task is too simple , I want more challenges 
 Data type : <class 'str'>


### <span style="color: red"> Task 2:</span> Create a function that finds the largest number in a list. Don not use built-in function. Show you function by using a random number generator to create the inputs.

### <span style="color: blue"> Solution:</span>

In [11]:
#importing dependencies
from random import randint

#defining function
def max_value(number_list):
    try:
        #take out first value and assign it as max val
        max_val = number_list.pop()
        #compares all values and assigns the largest to max val
        for value in number_list:
            if value > max_val:
                max_val = value
        return max_val
    except:
        return print("We could'nt find the max value. There might be somethinf")

#creating a list of random integers
random_list = [randint(1,100) for x in range(10)]

#printing out results
print(f"{random_list} \n max value of the list is {max_value(random_list)}")

[71, 25, 79, 2, 36, 76, 29, 39, 20, 52] 
 max value of the list is 79


### <span style="color: red"> Task 3:</span> Create a class named "Alloys". Initilize it with name and number of elements. In class "Alloys", write a method named "info", which prints "\<name> has \<num_of_elements> elements". For example, for Ti6Al4V, it will print "Ti6Al4V has 3 elements". Then create an inheritance calss of "Alloys" called "HEAs" (that's High Entropy Alloys). Create an instance from HEAs and print it's info. You can use CoCrFeMnNi, which has 5 elements.

### <span style="color: blue"> Solution:</span>

In [43]:
class Alloys:
    def __init__(self,name,num_of_elements):
        self.name = name
        self.elements = num_of_elements
    def info(self):
        print(f"{self.name} has {self.elements} elements")
class HEAs(Alloys):
    pass

HEA_instance = HEAs("CoCrFeMnNi",5)
HEA_instance.info()

CoCrFeMnNi has 5 elements


### <span style="color: red"> Task 4:</span> Write a function to check if an alloy’s composition is normalized. If it is not, your function should normalize it and return the normalized alloy name. E.g., if the input is $A_{1}B_{2}C_{3}$, your function should output A0.17B0.33C0.5. 
Bonus question: your function can run a check on the input name to check if there is any “not elements” in the name. Hint: you can try the Mendeleev python library for a complete list of elements.

### <span style="color: blue"> Solution:</span>

In [1]:
#importing dependencies
import re
from mendeleev import element

input_string = "ab134b0.52wowc31"
#pattern will recognize alternating character and numerical values ie A1B2C3 or A0.17B033C0.5
pattern = re.compile(r'([a-zA-Z]+)(\d+)')

#get all matching components of the string that match the regex
components = pattern.findall(input_string)

#Checking if the element is real
for component in components:
    try:
        if element(component[0]):
            print(f"{component[0]} is a real element")
    except:
        print(f"{component[0]} is a not a real element")

#summing all numerical values
total=0
for component in components:
    value = int(component[1])
    total += value

#Normalizing if the sum of the composition is not one

if total != 1:
    normalization_strings=[int(component[1])/total for component in components]
else:
    print("The formula given is already normalized")



#returning the formated string:
        



TypeError: 'tuple' object does not support item assignment

### <span style="color: red"> Task 5:</span> Some errors are presented in the code below. Can you identify and corret them?

NoResultFound: No row was found when one was required

In [ ]:
import os
import csv

current_dir = os.getcwd()
filename = 'data-1/E1_T1.csv' #adding this in case the file path needs to be changed
filename = current_dir + '/' + filename
with open(filename, 'r', newline='') as f:
    csv_file = csv.reader(f)
    data = next(csv_file)#changed to use next as the data is a single string
    
sentence=''

for word in data:
    sentence = sentence + ' '+ word

print(sentence)
print(type(sentence))

 This task is too simple , I want more challenges
<class 'str'>


### <span style="color: red"> Task 6:</span> Who has the most runner-up (i.e., second place) finishes in the histroy of NHL? Create a table with the teams and their runner-up finishes. The table must contain the the following columns:

-Team

-Total number of runner-up finishes

-Years [list of the years they finished as a runner-up]

You have to use Jupyter for the practice and produce a notebook with good readability. Data stored in 'Stanely_cup_winners_by_year.xlsx'.

### <span style="color: blue"> Solution:</span>

In [121]:
#first import dependencies
import pandas as pd
import numpy as np

#opening file
file = "data-1/Stanely_cup_winners_by_year.xlsx"
df = pd.read_excel(file)
df.head()

Year                 Team             Runner up
0  2020  Tampa Bay Lightning          Dallas Stars
1  2019      St. Louis Blues         Boston Bruins
2  2018  Washington Capitals  Vegas Golden Knights
3  2017  Pittsburgh Penguins   Nashville Predators
4  2016  Pittsburgh Penguins       San Jose Sharks

In [122]:
#dropping unncessesary columns
df.drop(['Team'],axis=1,inplace=True)
df.head()

Year             Runner up
0  2020          Dallas Stars
1  2019         Boston Bruins
2  2018  Vegas Golden Knights
3  2017   Nashville Predators
4  2016       San Jose Sharks

In [123]:
#creating a data frame that creates the list of years a team has won.
df1 = df.groupby('Runner up')['Year'].agg(list).reset_index()
df1.head()

Runner up                                               Year
0        Boston Bruins  [2019, 2013, 1990, 1988, 1978, 1977, 1974, 195...
1       Buffalo Sabres                                       [1999, 1975]
2       Calgary Flames                                       [2004, 1986]
3       Calgary Tigers                                             [1924]
4  Carolina Hurricanes                                             [2002]

In [129]:
#creating a data fram that has the total value counts of the times a team has been runner up
df2 = df['Runner up'].value_counts().rename_axis('Runner up').reset_index(name='Runner up count')
df2.head()

Runner up  Runner up count
0        Boston Bruins               14
1    Detroit Red Wings               13
2   Montreal Canadiens                9
3  Toronto Maple Leafs                8
4     New York Rangers                7

In [136]:
#merging both using the 'Runner up' column
df_final = pd.merge(df2,df1,on='Runner up',how='right')

In [137]:
#final table
df_final

Runner up  Runner up count  \
0             Boston Bruins               14   
1            Buffalo Sabres                2   
2            Calgary Flames                2   
3            Calgary Tigers                1   
4       Carolina Hurricanes                1   
5       Chicago Black Hawks                1   
6        Chicago Blackhawks                6   
7              Dallas Stars                2   
8         Detroit Red Wings               13   
9          Edmonton Eskimos                1   
10          Edmonton Oilers                2   
11         Florida Panthers                1   
12        Los Angeles Kings                1   
13  Mighty Ducks of Anaheim                1   
14    Minnesota North Stars                2   
15        Moncton Victories                1   
16       Montreal Canadiens                9   
17         Montreal Maroons                1   
18      Nashville Predators                1   
19        New Jersey Devils                2   
20       New York Islanders                1   
21         New York Rangers                7   
22          Ottawa Senators                2   
23      Philadelphia Flyers                6   
24      Pittsburgh Penguins                1   
25     Port Arthur Bearcats                1   
26        Portland Rosebuds                1   
27          San Jose Sharks                1   
28    Seattle Metropolitans                1   
29          St. Louis Blues                3   
30            Sydney Miners                1   
31      Tampa Bay Lightning                1   
32      Toronto Maple Leafs                8   
33        Vancouver Canucks                3   
34   Vancouver Millionaires                3   
35     Vegas Golden Knights                1   
36         Victoria Cougars                2   
37      Washington Capitals                1   

                                                 Year  
0   [2019, 2013, 1990, 1988, 1978, 1977, 1974, 195...  
1                                        [1999, 1975]  
2                                        [2004, 1986]  
3                                              [1924]  
4                                              [2002]  
5                                              [1931]  
6                [1992, 1973, 1971, 1965, 1962, 1944]  
7                                        [2020, 2000]  
8   [2009, 1995, 1966, 1964, 1963, 1961, 1956, 194...  
9                                              [1923]  
10                                       [2006, 1983]  
11                                             [1996]  
12                                             [1993]  
13                                             [2003]  
14                                       [1991, 1981]  
15                                             [1912]  
16  [1989, 1967, 1955, 1954, 1952, 1951, 1947, 192...  
17                                             [1928]  
18                                             [2017]  
19                                       [2012, 2001]  
20                                             [1984]  
21         [2014, 1979, 1972, 1950, 1937, 1932, 1929]  
22                                       [2007, 1915]  
23               [2010, 1997, 1987, 1985, 1980, 1976]  
24                                             [2008]  
25                                             [1911]  
26                                             [1916]  
27                                             [2016]  
28                                             [1920]  
29                                 [1970, 1969, 1968]  
30                                             [1913]  
31                                             [2015]  
32   [1960, 1959, 1940, 1939, 1938, 1936, 1935, 1933]  
33                                 [2011, 1994, 1982]  
34                                 [1922, 1921, 1918]  
35                                             [2018]  
36                                       [1926, 1914]  
37               

### <span style="color: red"> Task 7:</span> With the help of matminer, 1. Retrive the 'steel_strength' dataset into a dataframe and clean your dataset (i.e., do some basic data exploration and analysis work on your data. For instance, check the statistics of the dataset, check for NaNs, drop duplicated data entries etc. It is alway the first thing to do before you start working on a dataset.) if needed; 2. Featurize the compositions of the materials; and 3. Perform correlation analysis and drop feature(s) with a coefficient greater than abs(0.8) from your dataframe.

### <span style="color: blue"> Solution:</span>

In [1]:
#loading dependencies
from matminer.datasets import load_dataset
#creating the steel strength dataframe
df=load_dataset('steel_strength')
df.head()

ModuleNotFoundError: No module named 'matminer'

In [6]:
#statistical summary
df.describe()

c          mn          si          cr          ni          mo  \
count  312.000000  312.000000  312.000000  312.000000  312.000000  312.000000   
mean     0.096442    0.146250    0.221218    8.043830    8.184006    2.766090   
std      0.109008    0.397102    0.580796    5.426169    6.337055    1.832908   
min      0.000000    0.010000    0.010000    0.010000    0.010000    0.020000   
25%      0.010000    0.010000    0.010000    3.100000    0.960000    1.500000   
50%      0.030000    0.010000    0.010000    9.050000    8.500000    2.210000   
75%      0.182500    0.080000    0.110000   12.520000   12.117500    4.090000   
max      0.430000    3.000000    4.750000   17.500000   21.000000    9.670000   

                v           n          nb          co           w          al  \
count  312.000000  312.000000  312.000000  312.000000  312.000000  312.000000   
mean     0.183750    0.005545    0.035449    7.008782    0.161282    0.239135   
std      0.452462    0.018331    0.161537    6.254431    0.920211    0.340110   
min      0.000000    0.000000    0.000000    0.010000    0.000000    0.010000   
25%      0.010000    0.000000    0.010000    0.010000    0.000000    0.030000   
50%      0.010000    0.000000    0.010000    7.085000    0.000000    0.050000   
75%      0.127500    0.000000    0.010000   13.480000    0.000000    0.300000   
max      4.320000    0.150000    2.500000   20.100000    9.180000    1.800000   

               ti  yield strength  tensile strength  elongation  
count  312.000000      312.000000        312.000000  303.000000  
mean     0.310897     1420.998397       1641.653205   14.007261  
std      0.556593      301.894169        346.475179    5.095974  
min      0.000000     1005.900000       1019.000000    2.000000  
25%      0.000000     1219.475000       1338.125000   10.800000  
50%      0.030000     1344.200000       1666.300000   14.800000  
75%      0.232500     1576.075000       1899.950000   17.300000  
max      2.500000     2510.300000       2570.000000   35.000000

In [5]:
total_nan_count = df.isna().sum().sum()
total_nan_count

9

9 Missing values correspond with the statistical summary having 303 entries of elongation compared to the rest of the columns, all having 312 entries

In [9]:
#returns list of bools for each row if its duplicate or not
is_duplicate = df.duplicated()

#
for i,row in enumerate(is_duplicate):
    if row:
        print(f"{i} row is a duplicate")

In [10]:
from matminer.featurizers.composition import ElementFraction
element_fraction=ElementFraction()
element_fraction.feature_labels()

['H',
 'He',
 'Li',
 'Be',
 'B',
 'C',
 'N',
 'O',
 'F',
 'Ne',
 'Na',
 'Mg',
 'Al',
 'Si',
 'P',
 'S',
 'Cl',
 'Ar',
 'K',
 'Ca',
 'Sc',
 'Ti',
 'V',
 'Cr',
 'Mn',
 'Fe',
 'Co',
 'Ni',
 'Cu',
 'Zn',
 'Ga',
 'Ge',
 'As',
 'Se',
 'Br',
 'Kr',
 'Rb',
 'Sr',
 'Y',
 'Zr',
 'Nb',
 'Mo',
 'Tc',
 'Ru',
 'Rh',
 'Pd',
 'Ag',
 'Cd',
 'In',
 'Sn',
 'Sb',
 'Te',
 'I',
 'Xe',
 'Cs',
 'Ba',
 'La',
 'Ce',
 'Pr',
 'Nd',
 'Pm',
 'Sm',
 'Eu',
 'Gd',
 'Tb',
 'Dy',
 'Ho',
 'Er',
 'Tm',
 'Yb',
 'Lu',
 'Hf',
 'Ta',
 'W',
 'Re',
 'Os',
 'Ir',
 'Pt',
 'Au',
 'Hg',
 'Tl',
 'Pb',
 'Bi',
 'Po',
 'At',
 'Rn',
 'Fr',
 'Ra',
 'Ac',
 'Th',
 'Pa',
 'U',
 'Np',
 'Pu',
 'Am',
 'Cm',
 'Bk',
 'Cf',
 'Es',
 'Fm',
 'Md',
 'No',
 'Lr']

In [ ]:
correlation_matrix = df.corr()